File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string>res;
        backtrack(n,0,0,"",res);
        return res;
    }

    void backtrack(int n, int open,int close, string ans, vector<string>&res){
        if(open==n && close==n){
            res.push_back(ans);
        }
        if(open<n){
           backtrack(n,open+1,close,ans+'(',res);
        }
        if(close<open){
            backtrack(n,open,close+1,ans+')',res);
        }
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        res = []
        self.generate(n,0,0,"",res)
        return res
    def generate(self,n,open,close,str,res):
        if open==n and close==n:
            res.append(str)
        if open<n:
            self.generate(n,open+1,close,str+'(',res)
        if close<open:
            self.generate(n,open,close+1,str+')',res)
